cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1537,16 +1568,14 @@ class S3FileSystem::Impl {
       }
       req.SetBucket(ToAwsString(bucket));
       req.SetDelete(std::move(del));
-      delete_handlers.emplace_back();
-      futures.push_back(&delete_handlers.back().future);
-      client_->DeleteObjectsAsync(req, delete_handlers.back());
+      ARROW_ASSIGN_OR_RAISE(auto fut, io_context_.executor()->Submit(
+                                          io_context_.stop_token(), [client, req]() {
+                                            return client->DeleteObjects(req);
+                                          }));
+      futures.push_back(std::move(fut).Then(delete_cb));
     }
 
-    WaitForAll(futures);
-    for (const auto* fut : futures) {
-      RETURN_NOT_OK(fut->status());
-    }
-    return Status::OK();
+    return AllComplete(futures).status();
```

Comment:
Blocking at the end sort of defeats the purpose but it can be a fine first step.  If I do something like this I'll rename the method DeleteObjectsAsync that returns `Future<>` and then add back in `DeleteObjects` which calls `DeleteObjectsAsync` and waits on the status.  This ensures backwards compatibility but makes it even easier to consume the async version in the future.

Reply:
Well, this is implementing a sync method. The FileSystem interface doesn't have an async deletion method. We may add one, but I don't think that's necessary for now.